price-estimator/src/orderbook.rs

```diff
@@ -1,35 +1,83 @@
-use core::orderbook::StableXOrderBookReading;
-use pricegraph::Pricegraph;
-use std::time::{Duration, SystemTime};
+use crate::solver_rounding_buffer;
+use core::{
+    models::TokenId, orderbook::StableXOrderBookReading,
+    price_estimation::price_source::PriceSource,
+};
+use pricegraph::{Pricegraph, TokenPair};
+use std::{
+    sync::Arc,
+    time::{Duration, SystemTime},
+};
 use tokio::sync::RwLock;
 
+pub type DynPriceSource = Box<dyn PriceSource + Send + Sync>;
+
 /// Access and update the pricegraph orderbook.
-pub struct Orderbook<T> {
-    orderbook_reading: T,
+pub struct Orderbook {
+    orderbook_reading: Arc<dyn StableXOrderBookReading>,
+    price_source: DynPriceSource,
+    all_tokens: Vec<TokenId>,
     pricegraph: RwLock<Pricegraph>,
+    rounding_buffer_factor: f64,
 }
 
-impl<T> Orderbook<T> {
-    pub fn new(orderbook_reading: T) -> Self {
+impl Orderbook {
+    pub fn new(
+        orderbook_reading: Arc<dyn StableXOrderBookReading>,
+        price_source: DynPriceSource,
+        all_tokens: Vec<TokenId>,
+        rounding_buffer_factor: f64,
+    ) -> Self {
+        assert!(rounding_buffer_factor.is_finite());
+        assert!(rounding_buffer_factor >= 0.0);
         Self {
             orderbook_reading,
+            price_source,
+            all_tokens,
             pricegraph: RwLock::new(Pricegraph::new(std::iter::empty())),
+            rounding_buffer_factor,
         }
     }
 
-    pub async fn get_pricegraph(&self) -> Pricegraph {
+    pub async fn pricegraph(&self) -> Pricegraph {
         self.pricegraph.read().await.clone()
     }
-}
 
-impl<T: StableXOrderBookReading> Orderbook<T> {
+    /// See solver_rounding_buffer::rounding_buffer. This is used to adjust amounts in api queries
+    /// to adapt to how the solver sees them.
+    /// If any price is not available `1` is used as a fallback.
+    pub async fn rounding_buffer(&self, token_pair: TokenPair) -> f64 {
+        let fee_token = TokenId(0);
+        let sell_token = TokenId(token_pair.sell);
+        let buy_token = TokenId(token_pair.buy);
+        let prices = self
+            .price_source
+            .get_prices(&[fee_token, sell_token, buy_token])
+            .await
+            .unwrap_or_default();
+        let get_price = |token_id| prices.get(&token_id).copied().unwrap_or(1) as f64;
+        solver_rounding_buffer::rounding_buffer(
+            get_price(fee_token),
+            get_price(sell_token),
+            get_price(buy_token),
+        ) * self.rounding_buffer_factor
+    }
+
     /// Recreate the pricegraph orderbook.
     pub async fn update(&self) -> anyhow::Result<()> {
-        let (account_state, orders) = self
+        let (mut account_state, mut orders) = self
             .orderbook_reading
             .get_auction_data(current_batch_id() as u32)
             .await?;
 
+        let token_prices = self.price_source.get_prices(&self.all_tokens).await?;
+        solver_rounding_buffer::apply_rounding_buffer(
```

Comment:
Why keep it around when it's not used? The question I now see related to this is that we are actually changing /markets because it is working on these adjusted orders which wasn't the case before.

Reply:
I don't have a strong feeling about this one, but my intuition would be that markets should just show the state of the current order book as is. With price estimates we are actually making a soft commitment that if you use these prices, you will get matched (thus we need to apply our internal rounding buffer). But I can also see how you could argue for consistency, thus applying it to both.

Reply:
> I don't have a strong feeling about this one, but my intuition would be that markets should just show the state of the current order book as is. 

I agree with this - it would be counter-intuitive for a user to place an order and see it in the market orderbook page with a different price. Also, not applying the rounding buffer would keep it consistent with how it works today (to my knowledge).